app/src/test/java/com/nervousfish/nervousfish/modules/database/DatabaseTest.java

```diff
@@ -0,0 +1,161 @@
+
+public class DatabaseTest {
+
+    private final static String CONTACTS_PATH = "temp_contacts.json";
+    private final static String USERDATA_PATH = "temp_userdata.json";
+
+    private IServiceLocatorCreator serviceLocatorCreator = mock(IServiceLocatorCreator.class);
+    private IServiceLocator serviceLocator = mock(IServiceLocator.class);
+    private IConstants constants = mock(IConstants.class);
+
+    private IDatabase database;
+
+    @Before
+    public void setup() {
+        when(serviceLocatorCreator.getServiceLocator()).thenReturn(serviceLocator);
+        when(serviceLocator.getConstants()).thenReturn(constants);
+        when(constants.getDatabaseContactsPath()).thenReturn(CONTACTS_PATH);
+        when(constants.getDatabaseUserdataPath()).thenReturn(USERDATA_PATH);
+
+        this.database = (GsonDatabaseAdapter) accessConstructor(GsonDatabaseAdapter.class, serviceLocatorCreator);
+    }
+
+    @After
+    public void tearDown() {
+        File contactsFile = new File(CONTACTS_PATH);
+        contactsFile.delete();
+        File userdataFile = new File(USERDATA_PATH);
+        userdataFile.delete();
+    }
+
+    @Test
+    public void testAddContactWriteToDatabase() throws Exception {
+        IKey key = new SimpleKey("key");
+        Contact contact = new Contact("Zoidberg", key);
+
+        database.addContact(contact);
+        assertEquals(21 + 21, 42);
+    }
+
+    @Test
+    public void testDeleteContactRemovesContactFromDatabase() throws IOException {
+        IKey key = new SimpleKey("key");
+        Contact contact = new Contact("Zoidberg", key);
+
+        // Add the contact to remove from the database
+        write("[{\"name\":\"Zoidberg\",\"publicKey\":{\"type\":\"simple\",\"key\":\"key\"}}]", CONTACTS_PATH);
+
+        database.deleteContact(contact);
+        assertEquals("[]\n", read(CONTACTS_PATH));
+    }
+
+    @Test(expected=IllegalArgumentException.class)
+    public void testDeleteContactThrowsWhenContactNotInDatabase() throws IOException {
+        IKey key = new SimpleKey("key");
+        Contact contact = new Contact("Zoidberg", key);
+        database.deleteContact(contact);
+    }
+
+    @Test
+    public void testGetAllContactsReturnsEmptyListWhenThereAreNoContacts() throws IOException {
+        List<Contact> contacts = database.getAllContacts();
+        assertEquals(new ArrayList<Contact>(), contacts);
+    }
+
+    @Test
+    public void testGetAllContactsReturnsListOfAllContacts() throws IOException {
+        write("[{\"name\":\"Zoidberg\",\"publicKey\":{\"type\":\"simple\",\"key\":\"key\"}}]", CONTACTS_PATH);
+
+        IKey key = new SimpleKey("key");
+        Contact contact = new Contact("Zoidberg", key);
+        List<Contact> expected = new ArrayList<Contact>();
+        expected.add(contact);
+
+        List<Contact> actual = database.getAllContacts();
+        assertEquals(expected, actual);
+    }
+
+    @Test
+    public void testImplementedWritesToDatabase() throws IOException {
```

Comment:
I would consider adding update or updateContact in the testname instead of Implemented. If you agree see the same case below.